### ISSUE TITLE
Permission Denied Error in the middle/end of training. 

### ISSUE THREAD

**Issue by glakhanp** [ultralytics/ultralytics#14558]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

I am currently trying to run a federated learning experiment with YOLOv8 using multiple clients. I am currently training all the clients and facilitating the training process using a for loop that trains each client, aggregates all the weights, and then further trains each client using the aggregated weights.

However, I am encountering an odd issue where the model training will fail during the middle of training after an arbitrary epoch. The error I get is:

`
File ~\.conda\envs\insect_train\Lib\site-packages\ultralytics\engine\model.py:674, in Model.train(self, trainer, **kwargs)
    671             pass
    673 self.trainer.hub_session = self.session  # attach optional HUB session
--> 674 self.trainer.train()
    675 # Update model and cfg after training
    676 if RANK in {-1, 0}:

File ~\.conda\envs\insect_train\Lib\site-packages\ultralytics\engine\trainer.py:199, in BaseTrainer.train(self)
    196         ddp_cleanup(self, str(file))
    198 else:
--> 199     self._do_train(world_size)

File ~\.conda\envs\insect_train\Lib\site-packages\ultralytics\engine\trainer.py:432, in BaseTrainer._do_train(self, world_size)
    430     # Save model
    431     if self.args.save or final_epoch:
--> 432         self.save_model()
    433         self.run_callbacks("on_model_save")
    435 # Scheduler

File ~\.conda\envs\insect_train\Lib\site-packages\ultralytics\engine\trainer.py:501, in BaseTrainer.save_model(self)
    498 serialized_ckpt = buffer.getvalue()  # get the serialized content to save
    500 # Save checkpoints
--> 501 self.last.write_bytes(serialized_ckpt)  # save last.pt
    502 if self.best_fitness == self.fitness:
    503     self.best.write_bytes(serialized_ckpt)  # save best.pt

File ~\.conda\envs\insect_train\Lib\pathlib.py:1067, in Path.write_bytes(self, data)
   1065 # type-check for the buffer interface before truncating the file
   1066 view = memoryview(data)
-> 1067 with self.open(mode='wb') as f:
   1068     return f.write(view)

File ~\.conda\envs\insect_train\Lib\pathlib.py:1044, in Path.open(self, mode, buffering, encoding, errors, newline)
   1042 if "b" not in mode:
   1043     encoding = io.text_encoding(encoding)
-> 1044 return io.open(self, mode, buffering, encoding, errors, newline)

PermissionError: [Errno 13] Permission denied: 'C:\\Users\\glq2b\\Downloads\\fl_test\\client2\\runs\\weights\\last.pt'
`

Is there any solution or reason this might be happening? 

I am especially confused as this happens mid training. I am not even close to maxxing out my ram or vram, and I have full permissions to access the file.

Thanks!

### Environment

Ultralytics YOLOv8.2.35  Python-3.11.9 torch-2.0.1+cu117 CUDA:0 (NVIDIA RTX A5000, 24564MiB)
Setup complete  (56 CPUs, 126.6 GB RAM, 857.2/953.3 GB disk)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
It can happen if that checkpoint is open in some other program hence preventing overwrite. If multiple processes are accessing and writing to this file, then you should prevent that.

**Comment by glakhanp** [comment]
I feel that if that were the case, the model would fail to load the weight file at the start. I don't see why it would randomly fail to access the file right in the middle of training after some arbitrary epoch. 

As far as I can tell there is no other process attempting to access the files. while the model is training. 

Is there any idea why this might be happening?

I can however look more into if some other process is trying to access the last.pt weight file.

**Comment by Y-T-G** [comment]
It's the `last.pt` file as per the traceback. You can read a file that's open in another program. But you can't write to it. The traceback shows that it is trying to write to the file, probably after completing an epoch.

**Comment by glenn-jocher** [comment]
Marking as invalid, this is a local filehandling issue and likely user error.

**Comment by Smonkey123** [comment]
Thank you for your great YOLO models (v5, v8, 11). I have met the same question, and I tried many times to retrain, but the same error will generate, before this, I have never met this type of question.
In any tries, the error epoch number is different, such as 42, 56, 100+, 250+.
I have confirmed that I have all rights of this folder， but I don't know why, I have found several similar questions by filter (Permission Error), pls check this question, thank you very much.

`      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
    154/300      1.36G     0.2573     0.2557     0.8265         75        640: 100%|██████████| 10/10 [00:01<00:00,  9.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00, 14.20it/s]
                   all         20        164      0.999          1      0.995      0.983
  0%|          | 0/10 [00:00<?, ?it/s]
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
    155/300      1.36G     0.2547     0.2582     0.8339         76        640: 100%|██████████| 10/10 [00:01<00:00,  9.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00, 15.26it/s]
                   all         20        164      0.999          1      0.995      0.982
Traceback (most recent call last):
  File "C:\Users\CNXIGAO13\Desktop\eOCR\table_panel\train.py", line 25, in <module>
    main()
  File "C:\Users\CNXIGAO13\Desktop\eOCR\table_panel\train.py", line 16, in main
    model.train(data=data_yaml_path,
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\model.py", line 802, in train
    self.trainer.train()
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\trainer.py", line 207, in train
    self._do_train(world_size)
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\trainer.py", line 445, in _do_train
    self.save_model()
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\trainer.py", line 536, in save_model
    self.last.write_bytes(serialized_ckpt)  # save last.pt
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\pathlib.py", line 1223, in write_bytes
    with self.open(mode='wb') as f:
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\pathlib.py", line 1200, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\pathlib.py", line 1054, in _opener
    return self._accessor.open(self, flags, mode)
PermissionError: [Errno 13] Permission denied: 'runs\\detect\\train\\exp7\\weights\\last.pt'`

**Comment by Smonkey123** [comment]
Thank you for your great YOLO models (v5, v8, 11). I have met the same question, and I tried many times to retrain, but the same error will generate, before this, I have never met this type of question.
In any tries, the error epoch number is different, such as 42, 56, 100+, 250+.
I have confirmed that I have all rights of this folder， but I don't know why, I have found several similar questions by filter (Permission Error), pls check this question, thank you very much.

`            Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
    154/300      1.36G     0.2573     0.2557     0.8265         75        640: 100%|██████████| 10/10 [00:01<00:00,  9.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00, 14.20it/s]
                   all         20        164      0.999          1      0.995      0.983
  0%|          | 0/10 [00:00<?, ?it/s]
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
    155/300      1.36G     0.2547     0.2582     0.8339         76        640: 100%|██████████| 10/10 [00:01<00:00,  9.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00, 15.26it/s]
                   all         20        164      0.999          1      0.995      0.982
Traceback (most recent call last):
  File "C:\Users\CNXIGAO13\Desktop\eOCR\table_panel\train.py", line 25, in <module>
    main()
  File "C:\Users\CNXIGAO13\Desktop\eOCR\table_panel\train.py", line 16, in main
    model.train(data=data_yaml_path,
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\model.py", line 802, in train
    self.trainer.train()
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\trainer.py", line 207, in train
    self._do_train(world_size)
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\trainer.py", line 445, in _do_train
    self.save_model()
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\site-packages\ultralytics\engine\trainer.py", line 536, in save_model
    self.last.write_bytes(serialized_ckpt)  # save last.pt
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\pathlib.py", line 1223, in write_bytes
    with self.open(mode='wb') as f:
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\pathlib.py", line 1200, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "C:\ProgramData\anaconda3\envs\YOLO11\lib\pathlib.py", line 1054, in _opener
    return self._accessor.open(self, flags, mode)
PermissionError: [Errno 13] Permission denied: 'runs\\detect\\train\\exp7\\weights\\last.pt'`
![image](https://github.com/user-attachments/assets/675b39e1-95df-4aaf-8692-a9e5a127b4e2)

**Comment by Y-T-G** [comment]
@Smonkey123 Try excluding the folder from your antivirus or Windows Defender.

**Comment by Smonkey123** [comment]
Sorry to interrupt you again, I disabled the Windows Defender, this question continue exists. I found the weights (last.pt) wrote into this folder successfully (runs\\detect\\train\\exp7\\weights\\) when every time this error happens, so, I can't confirm that: 1, this file (last.pt) can't be overrode, 2, this path/folder can't be wrote.
![image](https://github.com/user-attachments/assets/68361e29-b5a2-4619-aed1-c43f704c5c15)

**Comment by Smonkey123** [comment]
@Y-T-G I continue to watch the weights folder in train process, the weight file (last.pt) keep updating (the [Date modified attribute changes as train process]), but up to some epoch, this error will appear.
![image](https://github.com/user-attachments/assets/bcaa6eed-d446-49fd-9def-c0d2e0a701bd)

**Comment by Smonkey123** [comment]
I found similar questions (#7087, #7509), and the real reason might be that Windows Defender is occupying/scanning last.pt. Unfortunately, my PC belongs to my company, and the firewall is controlled by the administrator. Another thing to mention is that I trained a 1-class detection model the day before yesterday, and everything went well. However, now I’m training a 2-class detection model and added ‘save_period=10’, but the same issue appears.
![image](https://github.com/user-attachments/assets/7e54f9c2-1546-47c7-a748-0ce943a2fa78)
